### PR TITLE
Fix jwtf support for ES* algorithms

### PR DESCRIPTION
**Pull request by rnewson**
ES256 support is apparently broken. Here's a test to prove it. I don't know the fix, though. I tried the same example (from jwt.io) with another library (https://github.com/G-Corp/jwerl) and it also fails. Is... Is erlang the problem?

### CODE REVIEW

**Comment by rnewson** [comment]
nope. turns out both jwerl and jwtf are wrong, the `ES*` signatures have a different format and need converting to and from. 

Except jwtf is now fixed.

File: src/jwtf/src/jwtf.erl [hunk 1]
```diff
@@ -70,7 +72,13 @@ encode(Header = {HeaderProps}, Claims, Key) ->
...
             case verification_algorithm(Alg) of
                 {public_key, Algorithm} ->
-                    public_key:sign(Message, Algorithm, Key);
+                    Signature = public_key:sign(Message, Algorithm, Key),
+                    case Alg of
+                        <<"ES", _/binary>> ->
+                            der_to_jose(Alg, Signature);
+                        _ ->
+                            Signature
+                    end;
```
Comment (jaydoane) [inline src/jwtf/src/jwtf.erl:81]: Can we factor this into a separate function and use it on lines 279-285 as well?

File: src/jwtf/src/jwtf.erl [hunk 2]
```diff
@@ -292,6 +307,24 @@ hmac_verify(Algorithm, Message, HMAC, SecretKey) ->
...
+    #'ECDSA-Sig-Value'{r = R, s = S} = public_key:der_decode('ECDSA-Sig-Value', Signature),
+    Len = rs_len(Alg),
+    <<R:Len, S:Len>>.
+
+rs_len(<<"ES256">>) ->
+    256;
+rs_len(<<"ES384">>) ->
+    384;
+rs_len(<<"ES512">>) ->
+    512.
```
Comment (jaydoane) [inline src/jwtf/src/jwtf.erl:326]: Is there a reason to prefer this implementation over e.g.
```erlang
rs_len(<<"ES", SizeBin/binary>>) ->
    binary_to_integer(SizeBin).

```

**Review by jaydoane — APPROVED** [review]

File: src/jwtf/src/jwtf.erl (hunk 2 above)
Comment (rnewson) [reply src/jwtf/src/jwtf.erl:326]: I like your version better.

File: src/jwtf/src/jwtf.erl (hunk 1 above)
Comment (rnewson) [reply src/jwtf/src/jwtf.erl:81]: I'm not following. lines 279-285 do the inverse operation. what are you suggesting?

File: src/jwtf/src/jwtf.erl [hunk 3]
```diff
@@ -268,12 +276,19 @@ key(Props, Checks, KS) ->
 verify(Alg, Header, Payload, SignatureOrMac0, Key) ->
     Message = <<Header/binary, $., Payload/binary>>,
     SignatureOrMac1 = b64url:decode(SignatureOrMac0),
+    SignatureOrMac2 =
+        case Alg of
+            <<"ES", _/binary>> ->
+                jose_to_der(SignatureOrMac1);
+            _ ->
+                SignatureOrMac1
+        end,
```
[the remark this answers is not in the document]
Comment (rnewson) [reply src/jwtf/src/jwtf.erl:285]: ?

File: src/jwtf/src/jwtf.erl (hunk 1 above)
Comment (jaydoane) [reply src/jwtf/src/jwtf.erl:97]: My bad, I failed to notice this one uses `der_to_jose` while the other `jose_to_der`. All I can say is "derp derp derp"

**Pull request merged**